File: kattis/brexit.py
from collections import deque

C, P, X, L = map(int, input().split())

graph = [[] for _ in range(C+1)]
thresh = [0]*(C+1)

for i in range(P):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
    thresh[a]+=1
    thresh[b]+=1

left = [0]*(C+1)
explored = set([L])
queue = deque()
queue.append(L)

while len(queue) > 0:
    curr = queue.popleft()
    for i in graph[curr]:
        if i not in explored:
            left[i]+=1
            if 2*left[i] >= thresh[i]:
                explored.add(i)
                queue.append(i)
            
if X in explored:
    print('leave')
else:
    print('stay')

File: kattis/brexit.cpp
#include <iostream>
#include <list>


using namespace std;


int Partners[200001][2], Out[200001]; //0-start,1-lost
list<int> Graph[200001];


void bfs(int s){
    list<int> queue;
    queue.push_back(s);
    Out[s] = 1;

    while (!queue.empty()){
        int curr = queue.front();
        queue.pop_front();

        for (auto neigh : Graph[curr]){
            if (Out[neigh]) continue;
            Partners[neigh][1]++;
            if (Partners[neigh][1]*2 >= Partners[neigh][0]) queue.push_back(neigh), Out[neigh] = 1;
        }
    }
}

int main(){
    cin.tie(NULL);
    ios_base::sync_with_stdio(false);
    int n, p, h, f;

    cin >> n >> p >> h >> f;

    for (int i=0;i<p;i++){
        int a,b;

        cin >> a >> b;
        
        Graph[a].push_back(b);
        Graph[b].push_back(a);

        Partners[a][0]++;
        Partners[b][0]++;
    }

    bfs(f);

    if (Out[h]){
        cout << "leave\n";
    }else{
        cout << "stay\n";
    }

    return 0;
}